### PR TITLE
Make Properties and Block generation in tests polymorphic

### PR DESCRIPTION
**Pull request by polinavino**
This PR : 
- makes all three variables in LedgerState's `RewardUpdate` into `DeltaCoin` to allow them to be negative
- in the `src` directory of the testing package, removes the constraint that the `era ~ ShelleyEra c`, and all the constraints that imply this from the `ShelleyTest` constraint bundle
- leaves most code in `scr` polymorphic *except* the actual generators of the transactions (`genTx`) so that most of it is re-usable for ShelleyMA
- introduces the class `GenFuncTx era` which contains a map `genFuncTx`, which, given relevant context, produces a transaction generator for transactions of a specific era
- allows block generation to be polymorphic by using the `genFuncTx` class function instead of `genTx` 
- Properties are polymorphic 
- Examples are constrained to `ShelleyEra c`

### CODE REVIEW

**Review by uroboros — CHANGES_REQUESTED** [review]
you'll have to separate the DeltaCoin work and generalisation of Examples before we can merge this - I suggest you wait for my era generalisation of the generators to be merged first...

**Comment by polinavino** [comment]
#1985 replaces this PR

**Pull request closed without merging**
